File: src/realtime.rs
// Realtime 模块 - 实时数据流处理
// 实时数据流处理系统

use std::collections::{HashMap, VecDeque};
use std::sync::{Arc, Mutex, RwLock, mpsc};
use std::thread;
use std::time::{Duration, Instant};
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use serde::{Deserialize, Serialize};
use crate::core::*;

// 流数据类型
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub enum StreamDataType {
    AgentState,
    Memory,
    Document,
    Vector,
    Event,
    Metric,
}

// 流数据项
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StreamDataItem {
    pub agent_id: u64,
    pub data_type: StreamDataType,
    pub payload: Vec<u8>,
    pub metadata: HashMap<String, String>,
    pub timestamp: i64,
    pub priority: u8,
}

impl StreamDataItem {
    pub fn new(agent_id: u64, data_type: StreamDataType, payload: Vec<u8>, metadata: HashMap<String, String>) -> Self {
        Self {
            agent_id,
            data_type,
            payload,
            metadata,
            timestamp: chrono::Utc::now().timestamp(),
            priority: 128, // 默认中等优先级
        }
    }

    pub fn with_priority(mut self, priority: u8) -> Self {
        self.priority = priority;
        self
    }

    pub fn is_high_priority(&self) -> bool {
        self.priority > 200
    }
}

// 流处理配置
#[derive(Debug, Clone)]
pub struct StreamProcessingConfig {
    pub buffer_size: usize,
    pub batch_size: usize,
    pub worker_threads: usize,
    pub batch_timeout: Duration,
    pub max_retries: u32,
}

impl Default for StreamProcessingConfig {
    fn default() -> Self {
        Self {
            buffer_size: 10000,
            batch_size: 100,
            worker_threads: 4,
            batch_timeout: Duration::from_millis(100),
            max_retries: 3,
        }
    }
}

// 流处理统计
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StreamProcessingStats {
    pub items_received: u64,
    pub items_processed: u64,
    pub items_dropped: u64,
    pub batches_processed: u64,
    pub avg_latency_ms: f64,
    pub max_latency_ms: u64,
    pub throughput_per_sec: f64,
    pub buffer_utilization: f64,
    pub error_count: u64,
    pub last_update: i64,
}

impl Default for StreamProcessingStats {
    fn default() -> Self {
        Self {
            items_received: 0,
            items_processed: 0,
            items_dropped: 0,
            batches_processed: 0,
            avg_latency_ms: 0.0,
            max_latency_ms: 0,
            throughput_per_sec: 0.0,
            buffer_utilization: 0.0,
            error_count: 0,
            last_update: chrono::Utc::now().timestamp(),
        }
    }
}

// 流数据处理器特征
pub trait StreamProcessor: Send + Sync {
    fn process_item(&self, item: &StreamDataItem) -> Result<(), Box<dyn std::error::Error + Send + Sync>>;
    fn process_batch(&self, items: &[StreamDataItem]) -> Result<(), Box<dyn std::error::Error + Send + Sync>>;
    fn get_processor_name(&self) -> &str;
}

// 实时流处理器
pub struct RealTimeStreamProcessor {
    config: StreamProcessingConfig,
    processors: HashMap<StreamDataType, Arc<dyn StreamProcessor>>,
    sender: mpsc::Sender<StreamDataItem>,
    receiver: Arc<Mutex<mpsc::Receiver<StreamDataItem>>>,
    buffer: Arc<Mutex<VecDeque<StreamDataItem>>>,
    stats: Arc<RwLock<StreamProcessingStats>>,
    is_running: Arc<AtomicBool>,
    worker_handles: Vec<thread::JoinHandle<()>>,
}

impl RealTimeStreamProcessor {
    pub fn new(config: StreamProcessingConfig) -> Self {
        let (sender, receiver) = mpsc::channel();
        
        Self {
            config,
            processors: HashMap::new(),
            sender,
            receiver: Arc::new(Mutex::new(receiver)),
            buffer: Arc::new(Mutex::new(VecDeque::new())),
            stats: Arc::new(RwLock::new(StreamProcessingStats::default())),
            is_running: Arc::new(AtomicBool::new(false)),
            worker_handles: Vec::new(),
        }
    }

    pub fn register_processor(&mut self, data_type: StreamDataType, processor: Arc<dyn StreamProcessor>) {
        self.processors.insert(data_type, processor);
    }

    pub fn start(&mut self) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        if self.is_running.load(Ordering::SeqCst) {
            return Err("Stream processor already running".into());
        }

        self.is_running.store(true, Ordering::SeqCst);

        // 启动工作线程
        for worker_id in 0..self.config.worker_threads {
            let receiver = Arc::clone(&self.receiver);
            let buffer = Arc::clone(&self.buffer);
            let stats = Arc::clone(&self.stats);
            let is_running = Arc::clone(&self.is_running);
            let processors = self.processors.clone();
            let config = self.config.clone();

            let handle = thread::spawn(move || {
                Self::worker_thread(worker_id, receiver, buffer, stats, is_running, processors, config);
            });

            self.worker_handles.push(handle);
        }

        println!("Real-time stream processor started with {} workers", self.config.worker_threads);
        Ok(())
    }

    pub fn stop(&mut self) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        self.is_running.store(false, Ordering::SeqCst);

        // 等待所有工作线程完成
        while let Some(handle) = self.worker_handles.pop() {
            let _ = handle.join();
        }

        Ok(())
    }

    pub fn submit_data(&self, item: StreamDataItem) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        if !self.is_running.load(Ordering::SeqCst) {
            return Err("Stream processor not running".into());
        }

        self.sender.send(item)?;

        // 更新统计
        if let Ok(mut stats) = self.stats.write() {
            stats.items_received += 1;
        }

        Ok(())
    }

    fn worker_thread(
        worker_id: usize,
        receiver: Arc<Mutex<mpsc::Receiver<StreamDataItem>>>,
        buffer: Arc<Mutex<VecDeque<StreamDataItem>>>,
        stats: Arc<RwLock<StreamProcessingStats>>,
        is_running: Arc<AtomicBool>,
        processors: HashMap<StreamDataType, Arc<dyn StreamProcessor>>,
        config: StreamProcessingConfig,
    ) {
        println!("Worker thread {} started", worker_id);

        while is_running.load(Ordering::SeqCst) {
            // 尝试接收数据
            if let Ok(receiver) = receiver.lock() {
                match receiver.try_recv() {
                    Ok(item) => {
                        let start_time = Instant::now();
                        
                        // 处理数据项
                        if let Some(processor) = processors.get(&item.data_type) {
                            match processor.process_item(&item) {
                                Ok(_) => {
                                    // 更新成功统计
                                    if let Ok(mut stats) = stats.write() {
                                        stats.items_processed += 1;
                                        let latency = start_time.elapsed().as_millis() as u64;
                                        if latency > stats.max_latency_ms {
                                            stats.max_latency_ms = latency;
                                        }
                                        // 更新平均延迟
                                        stats.avg_latency_ms = (stats.avg_latency_ms * 0.9) + (latency as f64 * 0.1);
                                    }
                                }
                                Err(_) => {
                                    // 更新错误统计
                                    if let Ok(mut stats) = stats.write() {
                                        stats.error_count += 1;
                                    }
                                }
                            }
                        } else {
                            // 没有对应的处理器，丢弃数据
                            if let Ok(mut stats) = stats.write() {
                                stats.items_dropped += 1;
                            }
                        }
                    }
                    Err(mpsc::TryRecvError::Empty) => {
                        // 没有数据，短暂休眠
                        thread::sleep(Duration::from_millis(1));
                    }
                    Err(mpsc::TryRecvError::Disconnected) => {
                        break;
                    }
                }
            }
        }

        println!("Worker thread {} stopped", worker_id);
    }

    pub fn get_stats(&self) -> StreamProcessingStats {
        self.stats.read().unwrap().clone()
    }

    pub fn get_buffer_size(&self) -> usize {
        self.buffer.lock().unwrap().len()
    }
}

// 流查询类型
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum StreamQueryType {
    VectorSimilarity,
    MemorySearch,
    AgentStateMonitor,
    EventPattern,
    RealTimeStats,
}

// 流查询
#[derive(Debug, Clone)]
pub struct StreamQuery {
    pub id: String,
    pub query_type: StreamQueryType,
    pub parameters: HashMap<String, String>,
    pub callback: String,
    pub created_at: Instant,
    pub last_result: Option<Vec<u8>>,
}

// 流式查询处理器
pub struct StreamQueryProcessor {
    query_cache: Arc<RwLock<HashMap<String, (Vec<u8>, Instant)>>>,
    cache_ttl: Duration,
    active_queries: Arc<RwLock<HashMap<String, StreamQuery>>>,
}

impl StreamQueryProcessor {
    pub fn new() -> Self {
        Self {
            query_cache: Arc::new(RwLock::new(HashMap::new())),
            cache_ttl: Duration::from_secs(300), // 5分钟缓存
            active_queries: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    pub fn register_query(&self, query: StreamQuery) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let mut queries = self.active_queries.write().unwrap();
        queries.insert(query.id.clone(), query);
        Ok(())
    }

    pub fn unregister_query(&self, query_id: &str) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let mut queries = self.active_queries.write().unwrap();
        queries.remove(query_id);
        Ok(())
    }

    pub fn process_stream_item(&self, item: &StreamDataItem) -> Result<Vec<String>, Box<dyn std::error::Error + Send + Sync>> {
        let mut triggered_queries = Vec::new();
        let queries = self.active_queries.read().unwrap();

        for (query_id, query) in queries.iter() {
            if self.should_trigger_query(query, item) {
                triggered_queries.push(query_id.clone());
            }
        }

        Ok(triggered_queries)
    }

    fn should_trigger_query(&self, query: &StreamQuery, item: &StreamDataItem) -> bool {
        match query.query_type {
            StreamQueryType::AgentStateMonitor => {
                if let Some(target_agent) = query.parameters.get("agent_id") {
                    if let Ok(agent_id) = target_agent.parse::<u64>() {
                        return item.agent_id == agent_id && 
                               matches!(item.data_type, StreamDataType::AgentState);
                    }
                }
                false
            }
            StreamQueryType::EventPattern => {
                matches!(item.data_type, StreamDataType::Event)
            }
            StreamQueryType::VectorSimilarity => {
                matches!(item.data_type, StreamDataType::Vector)
            }
            StreamQueryType::MemorySearch => {
                matches!(item.data_type, StreamDataType::Memory)
            }
            StreamQueryType::RealTimeStats => {
                true // 所有数据都可能触发统计查询
            }
        }
    }

    pub fn get_active_query_count(&self) -> usize {
        self.active_queries.read().unwrap().len()
    }

    pub fn get_cache_size(&self) -> usize {
        self.query_cache.read().unwrap().len()
    }
}

File: agent-db-zig/src/realtime.zig
// 实时数据流处理系统 - Zig API层
const std = @import("std");
const c = @cImport({
    @cInclude("agent_state_db.h");
});

// 流数据类型枚举
pub const StreamDataType = enum(u32) {
    AgentState = 0,
    Memory = 1,
    Document = 2,
    Vector = 3,
    Event = 4,
    Metric = 5,

    pub fn toString(self: StreamDataType) []const u8 {
        return switch (self) {
            .AgentState => "AgentState",
            .Memory => "Memory",
            .Document => "Document",
            .Vector => "Vector",
            .Event => "Event",
            .Metric => "Metric",
        };
    }
};

// 流查询类型枚举
pub const StreamQueryType = enum(u32) {
    VectorSimilarity = 0,
    MemorySearch = 1,
    AgentStateMonitor = 2,
    EventPattern = 3,
    RealTimeStats = 4,

    pub fn toString(self: StreamQueryType) []const u8 {
        return switch (self) {
            .VectorSimilarity => "VectorSimilarity",
            .MemorySearch => "MemorySearch",
            .AgentStateMonitor => "AgentStateMonitor",
            .EventPattern => "EventPattern",
            .RealTimeStats => "RealTimeStats",
        };
    }
};

// 流数据项结构
pub const StreamDataItem = struct {
    agent_id: u64,
    data_type: StreamDataType,
    payload: []const u8,
    priority: u8,

    pub fn init(agent_id: u64, data_type: StreamDataType, payload: []const u8) StreamDataItem {
        return StreamDataItem{
            .agent_id = agent_id,
            .data_type = data_type,
            .payload = payload,
            .priority = 128, // 默认中等优先级
        };
    }

    pub fn withPriority(self: StreamDataItem, priority: u8) StreamDataItem {
        var item = self;
        item.priority = priority;
        return item;
    }

    pub fn isHighPriority(self: StreamDataItem) bool {
        return self.priority > 200;
    }

    pub fn display(self: StreamDataItem) void {
        std.debug.print("Stream Data Item:\n", .{});
        std.debug.print("  Agent ID: {d}\n", .{self.agent_id});
        std.debug.print("  Data Type: {s}\n", .{self.data_type.toString()});
        std.debug.print("  Payload Size: {d} bytes\n", .{self.payload.len});
        std.debug.print("  Priority: {d}\n", .{self.priority});
    }
};

// 流处理统计结构
pub const StreamProcessingStats = struct {
    items_received: u64,
    items_processed: u64,
    items_dropped: u64,
    batches_processed: u64,
    avg_latency_ms: f64,
    max_latency_ms: u64,
    throughput_per_sec: f64,
    buffer_utilization: f64,
    error_count: u64,
    last_update: i64,

    pub fn display(self: StreamProcessingStats) void {
        std.debug.print("Stream Processing Statistics:\n", .{});
        std.debug.print("  Items Received: {d}\n", .{self.items_received});
        std.debug.print("  Items Processed: {d}\n", .{self.items_processed});
        std.debug.print("  Items Dropped: {d}\n", .{self.items_dropped});
        std.debug.print("  Batches Processed: {d}\n", .{self.batches_processed});
        std.debug.print("  Avg Latency: {d:.2} ms\n", .{self.avg_latency_ms});
        std.debug.print("  Max Latency: {d} ms\n", .{self.max_latency_ms});
        std.debug.print("  Throughput: {d:.2} items/sec\n", .{self.throughput_per_sec});
        std.debug.print("  Buffer Utilization: {d:.2}%\n", .{self.buffer_utilization * 100.0});
        std.debug.print("  Error Count: {d}\n", .{self.error_count});
    }

    pub fn getProcessingRate(self: StreamProcessingStats) f64 {
        if (self.items_received == 0) return 0.0;
        return @as(f64, @floatFromInt(self.items_processed)) / @as(f64, @floatFromInt(self.items_received));
    }

    pub fn getDropRate(self: StreamProcessingStats) f64 {
        if (self.items_received == 0) return 0.0;
        return @as(f64, @floatFromInt(self.items_dropped)) / @as(f64, @floatFromInt(self.items_received));
    }
};

// 实时流处理器
pub const RealTimeStreamProcessor = struct {
    processor: *c.CRealTimeStreamProcessor,
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator) !RealTimeStreamProcessor {
        const processor = c.stream_processor_new();
        if (processor == null) {
            return error.InitializationFailed;
        }

        return RealTimeStreamProcessor{
            .processor = processor.?,
            .allocator = allocator,
        };
    }

    pub fn deinit(self: *RealTimeStreamProcessor) void {
        c.stream_processor_free(self.processor);
    }

    pub fn start(self: *RealTimeStreamProcessor) !void {
        const result = c.stream_processor_start(self.processor);
        if (result != 0) {
            return error.StartFailed;
        }
    }

    pub fn stop(self: *RealTimeStreamProcessor) !void {
        const result = c.stream_processor_stop(self.processor);
        if (result != 0) {
            return error.StopFailed;
        }
    }

    pub fn submitData(self: *RealTimeStreamProcessor, item: StreamDataItem) !void {
        const result = c.stream_processor_submit_data(
            self.processor,
            item.agent_id,
            @intFromEnum(item.data_type),
            item.payload.ptr,
            item.payload.len,
            item.priority,
        );

        if (result != 0) {
            return error.SubmitDataFailed;
        }
    }

    pub fn getStats(self: *RealTimeStreamProcessor) !StreamProcessingStats {
        var stats: c.StreamProcessingStats = undefined;
        const result = c.stream_processor_get_stats(self.processor, &stats);
        if (result != 0) {
            return error.GetStatsFailed;
        }

        return StreamProcessingStats{
            .items_received = stats.items_received,
            .items_processed = stats.items_processed,
            .items_dropped = stats.items_dropped,
            .batches_processed = stats.batches_processed,
            .avg_latency_ms = stats.avg_latency_ms,
            .max_latency_ms = stats.max_latency_ms,
            .throughput_per_sec = stats.throughput_per_sec,
            .buffer_utilization = stats.buffer_utilization,
            .error_count = stats.error_count,
            .last_update = stats.last_update,
        };
    }

    pub fn getBufferSize(self: *RealTimeStreamProcessor) !u32 {
        const result = c.stream_processor_get_buffer_size(self.processor);
        if (result < 0) {
            return error.GetBufferSizeFailed;
        }
        return @intCast(result);
    }
};

// 流查询处理器
pub const StreamQueryProcessor = struct {
    processor: *c.CStreamQueryProcessor,
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator) !StreamQueryProcessor {
        const processor = c.stream_query_processor_new();
        if (processor == null) {
            return error.InitializationFailed;
        }

        return StreamQueryProcessor{
            .processor = processor.?,
            .allocator = allocator,
        };
    }

    pub fn deinit(self: *StreamQueryProcessor) void {
        c.stream_query_processor_free(self.processor);
    }

    pub fn registerQuery(
        self: *StreamQueryProcessor,
        query_id: []const u8,
        query_type: StreamQueryType,
        callback: []const u8,
    ) !void {
        const c_query_id = try self.allocator.dupeZ(u8, query_id);
        defer self.allocator.free(c_query_id);

        const c_callback = try self.allocator.dupeZ(u8, callback);
        defer self.allocator.free(c_callback);

        const result = c.stream_query_register(
            self.processor,
            c_query_id.ptr,
            @intFromEnum(query_type),
            c_callback.ptr,
        );

        if (result != 0) {
            return error.RegisterQueryFailed;
        }
    }

    pub fn unregisterQuery(self: *StreamQueryProcessor, query_id: []const u8) !void {
        const c_query_id = try self.allocator.dupeZ(u8, query_id);
        defer self.allocator.free(c_query_id);

        const result = c.stream_query_unregister(self.processor, c_query_id.ptr);
        if (result != 0) {
            return error.UnregisterQueryFailed;
        }
    }

    pub fn getActiveQueryCount(self: *StreamQueryProcessor) !u32 {
        const result = c.stream_query_get_active_count(self.processor);
        if (result < 0) {
            return error.GetActiveCountFailed;
        }
        return @intCast(result);
    }
};

// 流数据生成器（用于测试）
pub const StreamDataGenerator = struct {
    allocator: std.mem.Allocator,
    agent_id_counter: u64,

    pub fn init(allocator: std.mem.Allocator) StreamDataGenerator {
        return StreamDataGenerator{
            .allocator = allocator,
            .agent_id_counter = 1000,
        };
    }

    pub fn generateAgentStateData(self: *StreamDataGenerator) !StreamDataItem {
        self.agent_id_counter += 1;

        const state_data = try std.fmt.allocPrint(self.allocator, "{{\"agent_id\":{d},\"state\":\"active\",\"timestamp\":{d}}}", .{ self.agent_id_counter, std.time.timestamp() });

        return StreamDataItem.init(
            self.agent_id_counter,
            .AgentState,
            state_data,
        );
    }

    pub fn generateMemoryData(self: *StreamDataGenerator) !StreamDataItem {
        self.agent_id_counter += 1;

        const memory_data = try std.fmt.allocPrint(self.allocator, "{{\"agent_id\":{d},\"content\":\"Test memory content\",\"importance\":0.8}}", .{self.agent_id_counter});

        return StreamDataItem.init(
            self.agent_id_counter,
            .Memory,
            memory_data,
        );
    }

    pub fn generateEventData(self: *StreamDataGenerator) !StreamDataItem {
        self.agent_id_counter += 1;

        const event_data = try std.fmt.allocPrint(self.allocator, "{{\"agent_id\":{d},\"event_type\":\"user_interaction\",\"data\":[1.0,2.0,3.0]}}", .{self.agent_id_counter});

        return StreamDataItem.init(
            self.agent_id_counter,
            .Event,
            event_data,
        );
    }

    pub fn generateVectorData(self: *StreamDataGenerator) !StreamDataItem {
        self.agent_id_counter += 1;

        const vector_data = try std.fmt.allocPrint(self.allocator, "[[0.1,0.2,0.3,0.4,0.5],{{\"type\":\"embedding\",\"model\":\"test\"}}]", .{});

        return StreamDataItem.init(
            self.agent_id_counter,
            .Vector,
            vector_data,
        );
    }

    pub fn freeData(self: *StreamDataGenerator, data: []const u8) void {
        self.allocator.free(data);
    }
};
